File: Sources/COCHelperCore/VillageDetailFlatRow.swift
import Foundation

/// Issue #212：村庄详情扁平 render row 元数据（轻量 ID，不含 View body）。
///
/// 稳定 ID = 投影/快照 ID（禁止 offset / UUID——跨筛选/分页不变）。
/// 渲染时通过 `groupID` / `instanceID` / `itemID` 回查当前 tick 的
/// `BuildingGroup` / `BuildingInstance` / `VillageItemState`。
public enum VillageDetailFlatRow: Identifiable, Hashable, Sendable {
    /// 非精制台 section 头部：标题 + 完成度（Panel）。
    case sectionHeader(groupID: String, stats: VillageCategoryCompletion?)
    /// 精制台整组：标题 + 完成度 + 表格保留单 Panel 外观。
    case craftTable(groupID: String, stats: VillageCategoryCompletion?)
    /// 组头卡：`BuildingGroupSummaryView` 汇总 + Start/Cancel/Adjust 动作行。
    case groupHeader(groupID: String)
    /// 单实例块：实例行 + 该记录自己的升级阶梯（leadingDivider = 组内非首个）。
    case instance(groupID: String, instanceID: String, leadingDivider: Bool)
    /// 旧列表行（issue #24 父子缩进平铺；leadingDivider = 非 section 内首行）。
    case legacy(itemID: String, groupID: String, indented: Bool, leadingDivider: Bool)

    public var id: String {
        switch self {
        case .sectionHeader(let groupID, _):
            return "section:\(groupID)"
        case .craftTable(let groupID, _):
            return "craft:\(groupID)"
        case .groupHeader(let groupID):
            return "groupHeader:\(groupID)"
        case .instance(let groupID, let instanceID, _):
            return "instance:\(groupID):\(instanceID)"
        case .legacy(let itemID, _, _, _):
            return "legacy:\(itemID)"
        }
    }
}

File: packages/domain/src/village/village-detail-flat-row.ts
import type { BuildingGroup } from './building-group-projection';
import type { VillageCategoryCompletion, VillageDetailGroup } from './village-detail-projection';
import { villageDetailParentedRows } from './village-detail-projection';

export type VillageDetailFlatRow =
  | {
      readonly kind: 'sectionHeader';
      readonly groupID: string;
      readonly stats: VillageCategoryCompletion | null | undefined;
    }
  | {
      readonly kind: 'craftTable';
      readonly groupID: string;
      readonly stats: VillageCategoryCompletion | null | undefined;
    }
  | {
      readonly kind: 'groupHeader';
      readonly groupID: string;
    }
  | {
      readonly kind: 'instance';
      readonly groupID: string;
      readonly instanceID: string;
      readonly leadingDivider: boolean;
    }
  | {
      readonly kind: 'legacy';
      readonly itemID: string;
      readonly groupID: string;
      readonly indented: boolean;
      readonly leadingDivider: boolean;
    };

export function villageDetailFlatRowId(row: VillageDetailFlatRow): string {
  switch (row.kind) {
    case 'sectionHeader':
      return `section:${row.groupID}`;
    case 'craftTable':
      return `craft:${row.groupID}`;
    case 'groupHeader':
      return `groupHeader:${row.groupID}`;
    case 'instance':
      return `instance:${row.groupID}:${row.instanceID}`;
    case 'legacy':
      return `legacy:${row.itemID}`;
  }
}

export function buildVillageDetailFlatRows(input: {
  readonly displayGroups: readonly VillageDetailGroup[];
  readonly statsByKey: Readonly<Record<string, VillageCategoryCompletion>>;
  readonly groupByInstanceID: Readonly<Record<string, BuildingGroup>>;
}): VillageDetailFlatRow[] {
  const rows: VillageDetailFlatRow[] = [];
  for (const group of input.displayGroups) {
    if (group.displayCategory === 'craftTable') {
      rows.push({
        kind: 'craftTable',
        groupID: group.id,
        stats: input.statsByKey[group.id],
      });
      continue;
    }

    rows.push({
      kind: 'sectionHeader',
      groupID: group.id,
      stats: input.statsByKey[group.id],
    });

    const orderedGroups: BuildingGroup[] = [];
    const seenGroupIDs = new Set<string>();
    const fallbackItems = [];
    for (const item of group.items) {
      const buildingGroup = input.groupByInstanceID[rawRecordID(item.id)];
      if (buildingGroup !== undefined) {
        if (!seenGroupIDs.has(buildingGroup.id)) {
          seenGroupIDs.add(buildingGroup.id);
          orderedGroups.push(buildingGroup);
        }
      } else {
        fallbackItems.push(item);
      }
    }

    for (const buildingGroup of orderedGroups) {
      rows.push({ kind: 'groupHeader', groupID: buildingGroup.id });
      buildingGroup.instances.forEach((instance, index) => {
        rows.push({
          kind: 'instance',
          groupID: buildingGroup.id,
          instanceID: instance.id,
          leadingDivider: index > 0,
        });
      });
    }

    const parented = villageDetailParentedRows(fallbackItems);
    parented.forEach((row, index) => {
      rows.push({
        kind: 'legacy',
        itemID: row.item.id,
        groupID: group.id,
        indented: false,
        leadingDivider: index > 0,
      });
      for (const child of row.children) {
        rows.push({
          kind: 'legacy',
          itemID: child.id,
          groupID: group.id,
          indented: true,
          leadingDivider: true,
        });
      }
    });
  }
  return rows;
}

export function groupBuildingGroupsByInstanceID(
  buildingGroups: readonly BuildingGroup[],
): Record<string, BuildingGroup> {
  const entries = buildingGroups.flatMap((group) =>
    group.instances.map((instance) => [instance.id, group] as const),
  );
  const result: Record<string, BuildingGroup> = {};
  for (const [instanceID, group] of entries) {
    if (result[instanceID] === undefined) {
      result[instanceID] = group;
    }
  }
  return result;
}

export function rawRecordID(id: string): string {
  return id.startsWith('agg:') ? id.slice(4) : id;
}

export const VillageDetailFlatRowProjection = {
  build: buildVillageDetailFlatRows,
  groupByInstanceID: groupBuildingGroupsByInstanceID,
  rawRecordID,
};
